roachtest: tpcc/w=100/nodes=3/chaos=true failed

SHA: https://github.com/cockroachdb/cockroach/commits/ca1ef4d4f8296b213c0b2b140f16e4a97931e6e7

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpcc/w=100/nodes=3/chaos=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1368144&tab=buildLog

```
The test failed on branch=master, cloud=gce:
	cluster.go:1511,tpcc.go:167,cluster.go:1849,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1368144-tpcc-w-100-nodes-3-chaos-true:4 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=30m0s {pgurl:1-3} returned:
		stderr:
		
		stdout:
		l
		   24m8s 11676216            0.0            3.1      0.0      0.0      0.0      0.0 delivery
		   24m8s 11676216            0.0           30.1      0.0      0.0      0.0      0.0 newOrder
		   24m8s 11676216            0.0            3.2      0.0      0.0      0.0      0.0 orderStatus
		   24m8s 11676216            0.0           29.1      0.0      0.0      0.0      0.0 payment
		   24m8s 11676216            0.0            3.1      0.0      0.0      0.0      0.0 stockLevel
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   24m9s 11676216            0.0            3.1      0.0      0.0      0.0      0.0 delivery
		   24m9s 11676216            0.0           30.1      0.0      0.0      0.0      0.0 newOrder
		   24m9s 11676216            0.0            3.2      0.0      0.0      0.0      0.0 orderStatus
		   24m9s 11676216            0.0           29.0      0.0      0.0      0.0      0.0 payment
		   24m9s 11676216            0.0            3.1      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1870,tpcc.go:177,tpcc.go:313,test.go:1249: unexpected node event: 2: dead
	cluster.go:1033,context.go:122,cluster.go:1022,panic.go:406,test.go:783,test.go:769,cluster.go:1870,tpcc.go:177,tpcc.go:313,test.go:1249: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1368144-tpcc-w-100-nodes-3-chaos-true --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		2: dead
		1: 9492
		3: 9436
		Error:  2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/4878cb3e960dc26f0946e527e1836f27b3304c97

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpcc/w=100/nodes=3/chaos=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1373334&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190704-1373334/tpcc/w=100/nodes=3/chaos=true/run_1
	cluster.go:1724,tpcc.go:173,cluster.go:2067,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1562220659-01-n4cpu4:4 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=30m0s {pgurl:1-3} returned:
		stderr:
		
		stdout:
		   0.0 newOrder
		   4m39s  2331525            0.0            6.6      0.0      0.0      0.0      0.0 orderStatus
		   4m39s  2331525            0.0           63.9      0.0      0.0      0.0      0.0 payment
		   4m39s  2331525            0.0            6.5      0.0      0.0      0.0      0.0 stockLevel
		E190704 06:49:17.299757 1 workload/cli/run.go:427  error in stockLevel: dial tcp 10.142.0.140:26257: connect: connection refused
		   4m40s  2386776            0.0            6.6      0.0      0.0      0.0      0.0 delivery
		   4m40s  2386776            0.0           64.9      0.0      0.0      0.0      0.0 newOrder
		   4m40s  2386776            0.0            6.5      0.0      0.0      0.0      0.0 orderStatus
		   4m40s  2386776            0.0           63.6      0.0      0.0      0.0      0.0 payment
		   4m40s  2386776            0.0            6.5      0.0      0.0      0.0      0.0 stockLevel
		E190704 06:49:18.299829 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.140:26257: connect: connection refused
		: signal: killed
	cluster.go:2088,tpcc.go:183,tpcc.go:319,test_runner.go:681: unexpected node event: 3: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/36b74f0373b040aa873d4c23b415d94deba88378

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpcc/w=100/nodes=3/chaos=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1376442&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190708-1376442/tpcc/w=100/nodes=3/chaos=true/run_1
	cluster.go:1724,tpcc.go:173,cluster.go:2067,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1562566386-08-n4cpu4:4 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=30m0s {pgurl:1-3} returned:
		stderr:
		
		stdout:
		     0.0      0.0 orderStatus
		   4m28s  4001350            0.0           43.4      0.0      0.0      0.0      0.0 payment
		   4m28s  4001350            0.0            4.7      0.0      0.0      0.0      0.0 stockLevel
		E190708 06:58:43.659806 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.103:26257: connect: connection refused
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   4m29s  4013678            0.0            4.6      0.0      0.0      0.0      0.0 delivery
		   4m29s  4013678            0.0           44.8      0.0      0.0      0.0      0.0 newOrder
		   4m29s  4013678            0.0            4.7      0.0      0.0      0.0      0.0 orderStatus
		   4m29s  4013678            0.0           43.2      0.0      0.0      0.0      0.0 payment
		   4m29s  4013678            0.0            4.6      0.0      0.0      0.0      0.0 stockLevel
		E190708 06:58:44.659871 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.103:26257: connect: connection refused
		: signal: killed
	cluster.go:2088,tpcc.go:183,tpcc.go:319,test_runner.go:681: unexpected node event: 3: dead

```

--- a/pkg/kv/txn_interceptor_committer.go
+++ b/pkg/kv/txn_interceptor_committer.go
@@ -106,7 +106,6 @@ var parallelCommitsEnabled = settings.RegisterBoolSetting(
 // If it is unknown whether all of the requests in the final batch succeeded
 // (e.g. due to a network error) then an AmbiguousResultError is returned. The
 // logic to enforce this is in DistSender.
-// TODO(nvanbenschoten): merge this logic.
 //
 // In all cases, the interceptor abstracts away the details of this from all
 // interceptors above it in the coordinator interceptor stack.

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -12,12 +12,14 @@ package roachpb
 
 import (
 	"bytes"
+	"context"
 	"errors"
 	"fmt"
 	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/util/caller"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 )
 
@@ -189,6 +191,7 @@ func (e *Error) SetDetail(err error) {
 				panic(fmt.Sprintf("transactionRestartError %T must be an ErrorDetail", err))
 			}
 		}
+		e.checkTxnStatusValid()
 	}
 }
 
@@ -215,6 +218,7 @@ func (e *Error) SetTxn(txn *Transaction) {
 		// Refresh the message as the txn is updated.
 		e.Message = sErr.message(e)
 	}
+	e.checkTxnStatusValid()
 }
 
 // GetTxn returns the txn.
@@ -234,6 +238,23 @@ func (e *Error) UpdateTxn(o *Transaction) {
 		e.UnexposedTxn = o.Clone()
 	} else {
 		e.UnexposedTxn.Update(o)
+	}
+	e.checkTxnStatusValid()
+}
+
+// checkTxnStatusValid verifies that the transaction status is in-sync with the
+// error detail.
+func (e *Error) checkTxnStatusValid() {
+	if e.UnexposedTxn == nil {
+		return
+	}
+	txnStatus := e.UnexposedTxn.Status
+	if r, ok := e.Detail.GetInner().(transactionRestartError); !ok {
+		return
+	} else if _, ok := r.(*TransactionAbortedError); !ok {
+		if txnStatus.IsFinalized() {
+			log.Fatalf(context.TODO(), "transaction unexpectedly finalized in (%T): %v", r, e)
+		}
 	}
 }
 
@@ -789,6 +810,11 @@ func (e *IntentMissingError) message(_ *Error) string {
 }
 
 var _ ErrorDetailInterface = &IntentMissingError{}
+var _ transactionRestartError = &IntentMissingError{}
+
+func (*IntentMissingError) canRestartTransaction() TransactionRestart {
+	return TransactionRestart_IMMEDIATE
+}
 
 func (e *MergeInProgressError) Error() string {
 	return e.message(nil)

